### PR TITLE
Tier 0 ownership enforcement: tenant/project IDOR checks (#42/#43/#54/#55/#73)

### LINKED ISSUE #42: Subscription bundle IDOR: get_bundle/delete accept arbitrary IDs with no ownership check
**Severity:** High  ·  **Area:** Security

**Location:** `src/core/subscriptions.py:36-53`; `src/core/mcp_adapter.py:57-62`

**Problem:** Both `get_bundle` and `delete` query by `subscription_id` alone (no tenant/project filter). IDs are 128-bit UUIDs but are broadcast on Redis channel `subscription:{id}:updated`, and in default config the MCP endpoint is unauthenticated so any HTTP client can read or delete any subscription. The defect survives auth being enabled — the fix must be in the service layer.

**Impact:** Any client (unauthenticated in default config) can read or delete any subscription by ID. IDs are leaked via Redis pub/sub channels.

**Fix:** Add `tenant_id`/`project_id` to the WHERE clause, sourced from authenticated state.

---
Source: 2026-08-22 adversarial audit (`docs/audits/2026-08-22-adversarial-audit.md`)

### PR DESCRIPTION
**Pull request by robmillersoftware**
## Description

**Tier 0 ownership enforcement (Wave D fast-follow, PR B of 2).** Builds on the merged always-a-tenant foundation (#130) to close the tenant/project IDOR gaps: a request can only touch projects, keys, and subscriptions that belong to the caller's tenant. Every check is a **no-op unless `MULTI_TENANT_ENABLED` is on and the caller carries a tenant** — single-tenant and auth-off/demo deployments are unaffected.

Design doc: `docs/superpowers/plans/2026-09-07-tier0-ownership-batch1.md`. Requires #130 (merged).

## Type of Change

- [x] Bug fix (closes IDOR/authorization gaps)
- [x] New feature (per-route/tool tenant ownership enforcement)

## Related Issues

Closes #42, #43, #54, #55, #73
Also removes the dead `require_admin_permission` guard left over from #39, and the sandbox data-path (#40 REST remnant).

## Changes Made

- **`check_project_access` helper** (`src/core/ownership.py`): gated project→tenant ownership — read paths deny unowned/cross-tenant (403); publish paths auto-bind a new project to the caller's tenant. Resolves ownership via `TenantManager.get_project_tenant` (the `tenant_projects` mapping).
- **#54 / #55 — data & cleanup routes** (`src/api/routes.py`): publish/upload/import/batch auto-bind; query/events/data and admin cleanup/retention deny cross-tenant.
- **#43 — API keys** (`src/api/routes.py`, `src/core/auth.py`): `GET /auth/keys` lists only the caller's tenant; `DELETE` refuses cross-tenant (404, no info leak); `POST` stamps the creator's tenant.
- **`/sandbox/subscribe`**: project-scope check before the SSE stream opens.
- **#42 / #73 — subscriptions** (`src/core/subscriptions.py`, `src/core/mcp_adapter.py`): `get_bundle`/`delete` verify the subscription's tenant (raise before returning/deleting on mismatch); the MCP tools (`contex_create_subscription`/`contex_delete_subscription`/`read_subscription`) stamp/verify tenant from the token claims, so the create→read→delete lifecycle is tenant-consistent.
- **Removed the dead `require_admin_permission`** (#39 remnant, ~28 sites across tenant/webhook/service-account/audit routers) — each route is already gated by `require(Permission.MANAGE_*/VIEW_*)`.

## Testing

- [x] All existing tests pass
- [x] New tests added for new functionality
- [x] Tested edge cases

Full `tests/` suite: **464 passed, 0 failed**. New: `tests/test_ownership.py`, `tests/test_subscription_ownership.py`, and `tests/test_tenant_isolation.py` — an end-to-end matrix proving cross-tenant project data → 403, key listing scoped, publish auto-bind, subscription cross-tenant denial, sandbox 403, **and** that with `MULTI_TENANT_ENABLED=false` the same cross-tenant calls are *not* denied (isolation is genuinely toggled, not always-on). A multi-tenant MCP create→read→delete lifecycle test guards the create-tenant-stamping path. (~6 `sdk/python/tests` failures are pre-existing/unrelated.)

## Checklist

- [x] Self-review + per-task review of each change; whole-branch review (caught and fixed a create-subscription tenant-stamping gap)
- [x] Docs updated (plan committed)
- [x] No new warnings
- [x] Tests added proving cross-tenant denial and the isolation-off no-op
- [x] Existing tests pass locally

## Additional Notes

**Deferred (follow-up, non-blocking):** the revoke-failure audit-log message conflates "not found" with cross-tenant mismatch (internal log only, no caller leak); a pre-existing unused `public` import in `tenant_routes.py`. RLS (#64/#68/#69) remains a separate later effort.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/core/ownership.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+"""Project→tenant ownership enforcement for authorized routes."""
+from __future__ import annotations
+
+from fastapi import HTTPException
+
+from src.core.identity import Identity
+from src.core.tenant_middleware import MULTI_TENANT_ENABLED
+
+
+async def check_project_access(
```
Comment (robmillersoftware) [inline src/core/ownership.py]: Do you think it would make more sense if this returned a boolean instead of raising exceptions or being a no-op?

File: src/api/routes.py [hunk 2]
```diff
@@ -28,6 +31,10 @@
 logger = get_logger(__name__)
 
 
+def get_tenant_manager(request: Request) -> TenantManager:
```
Comment (robmillersoftware) [inline src/api/routes.py]: Is there a reason this can't be dependency injected or only instantiated once?

File: src/api/routes.py [hunk 3]
```diff
@@ -364,8 +371,8 @@ async def list_permissions():
     }
 
 
-@router.post("/data/publish", response_model=dict, dependencies=[Depends(require(Permission.PUBLISH_DATA))])
```
Comment (robmillersoftware) [inline src/api/routes.py:367]: I would greatly prefer to leave the permissions checks in the annotation. Is there a reason we can't do that?

File: src/core/auth.py [hunk 4]
```diff
@@ -118,20 +118,36 @@ async def create_api_key(
...
+async def revoke_api_key(
+    db: DatabaseManager, key_id: str, tenant_id: Optional[str] = None
+) -> bool:
     """
     Revoke an API key by ID.
 
     Args:
         db: Database manager
         key_id: Key ID to revoke
+        tenant_id: When provided, only revokes the key if it belongs to this tenant
```
Comment (robmillersoftware) [inline src/core/auth.py:130]: Is there a reason we need this?

File: src/web/routes.py [hunk 5]
```diff
@@ -263,17 +266,20 @@ async def get_project_data(request: Request, project_id: str):
...
     project_id: str = Query(...),
     need: str = Query(...),
+    identity: Identity = Depends(require(Permission.QUERY_DATA)),
 ):
     """Stream a natural-language need as a live-updating context bundle over SSE.
 
     Backed by an ephemeral Subscription; the browser is a parallel consumer of the
     same reconcile pipeline the MCP bridge uses.
     """
+    tenant_mgr = TenantManager(request.app.state.db)
```
Comment (robmillersoftware) [inline src/web/routes.py]: See, the other file uses a helper method to construct this. This is already getting sloppy. Can we not use dependency injection?

File: src/core/subscriptions.py [hunk 6]
```diff
@@ -10,10 +10,16 @@
 
 from src.core.db_models import Subscription
 from src.core.tenant import DEFAULT_TENANT_ID
+from src.core.tenant_middleware import MULTI_TENANT_ENABLED
 
 logger = logging.getLogger(__name__)
 
 
+def _assert_sub_tenant(row, tenant_id):
```
Comment (robmillersoftware) [inline src/core/subscriptions.py:18]: Explain this one to me.

File: src/core/subscriptions.py [hunk 7]
```diff
@@ -34,22 +40,24 @@ async def create(
             await session.commit()
         return sub_id
 
-    async def get_bundle(self, subscription_id) -> dict:
+    async def get_bundle(self, subscription_id, *, tenant_id=None) -> dict:
```
Comment (robmillersoftware) [inline src/core/subscriptions.py:43]: Why the wildcard here?

File: src/core/ownership.py (hunk 1 above)
Comment (robmillersoftware) [reply src/core/ownership.py]: Done — `check_project_access` returns a bool now (True = allowed); the FastAPI `HTTPException` import is gone from `src/core/ownership.py` and the routes raise the 403 on False. (a4fe8ed)

File: src/api/routes.py (hunk 2 above)
Comment (robmillersoftware) [reply src/api/routes.py]: Consolidated to a single `get_tenant_manager` in `src/api/deps.py`, injected via `Depends(get_tenant_manager)`. The duplicate defs and the inline construction are gone. (7ce93bc)

File: src/web/routes.py (hunk 5 above)
Comment (robmillersoftware) [reply src/web/routes.py]: Fixed — this now uses the same shared `get_tenant_manager` dependency from `src/api/deps.py` (via `Depends(...)`) instead of the inline `TenantManager(...)`. One accessor everywhere. (7ce93bc)

File: src/api/routes.py (hunk 3 above)
Comment (robmillersoftware) [reply src/api/routes.py:367]: Restored — `require(Permission.X)` is back in `dependencies=[...]` on the decorators; handlers that need the caller get it via `identity: Identity = Depends(get_identity)`. FastAPI caches get_identity, so require() and the param resolve it once (no double lookup). (7ce93bc)

File: src/core/auth.py (hunk 4 above)
Comment (robmillersoftware) [reply src/core/auth.py:130]: The tenant scoping is #43 (block cross-tenant revoke), but the SELECT-then-DELETE was unnecessary — collapsed to a single `DELETE ... WHERE key_id AND tenant_id` with a rowcount check. (4018a09)

File: src/core/subscriptions.py (hunk 6 above)
Comment (robmillersoftware) [reply src/core/subscriptions.py:18]: It is the subscription-level ownership gate: raises `PermissionError` only when multi-tenancy is on AND the caller carries a tenant AND the subscription belongs to a different tenant; otherwise a no-op (single-tenant / demo / tenant-less callers untouched). It is the subscription analog of `check_project_access`, raising a plain builtin (not FastAPI) since it is service-layer and surfaced by the MCP tool. Left as-is; happy to make it a bool for symmetry if you prefer.

File: src/core/subscriptions.py (hunk 7 above)
Comment (robmillersoftware) [reply src/core/subscriptions.py:43]: The `*` makes `tenant_id` keyword-only, so callers must write `tenant_id=...` rather than pass it positionally (it is an ownership-scoping arg, not a normal positional). Purely a clarity guard — I can drop it if you find it noisy.

**Comment by robmillersoftware** [comment]
Also folded in a fix the re-review surfaced: `POST /batch/publish` was running the ownership check inside the per-item try/except (and the outer handler wraps everything in `except Exception -> 500`), so a cross-tenant project in a batch was swallowed into a per-item failure and returned 200. Now enforced up-front before the best-effort loop, so a cross-tenant batch returns 403 (with a regression test). (a84f9dd)

File: src/core/ownership.py (hunk 1 above)
Comment (robmillersoftware) [reply src/core/ownership.py]: Reworked per your follow-up: `check_project_access` is now `ensure_project_access`, which raises a domain `PermissionError` from core (no FastAPI import in core anymore), and a single app-level handler maps `PermissionError -> 403`. The ~9 per-route `if not: raise HTTPException` lines are gone — routes just `await ensure_project_access(...)`. (6d8bbd9)

File: src/core/subscriptions.py (hunk 6 above)
Comment (robmillersoftware) [reply src/core/subscriptions.py:18]: Now fully consistent: both `ensure_project_access` and `_assert_sub_tenant` raise a domain `PermissionError` from core, translated to 403 by the one app-level handler (the MCP path surfaces it as a tool error). Core no longer references FastAPI. (6d8bbd9)
